patchwork-capabilities/src/main/java/net/minecraftforge/event/AttachCapabilitiesEvent.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Minecraft Forge, Patchwork Project
+ * Copyright (c) 2016-2019, 2019
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.event;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import net.minecraftforge.common.capabilities.ICapabilityProvider;
+import net.minecraftforge.eventbus.api.GenericEvent;
+
+import net.minecraft.util.Identifier;
+
+/**
+ * Fired whenever an object with Capabilities support (currently {@link net.minecraft.block.entity.BlockEntity block entities}, {@link net.minecraft.item.Item items}, and {@link javax.swing.text.html.parser.Entity entities})
+ * is created. Allowing for the attachment of arbitrary capability providers.
+ *
+ * <p>Please note that as this is fired for ALL object creations efficient code is recommended.
+ * And if possible use one of the sub-classes to filter your intended objects.
+ */
+public class AttachCapabilitiesEvent<T> extends GenericEvent<T> {
+	private final T obj;
+	private final Map<Identifier, ICapabilityProvider> caps = Maps.newLinkedHashMap();
+	private final Map<Identifier, ICapabilityProvider> view = Collections.unmodifiableMap(caps);
+	private final List<Runnable> listeners = Lists.newArrayList();
+	private final List<Runnable> listenersView = Collections.unmodifiableList(listeners);
+
+	// For EventBus
+	public AttachCapabilitiesEvent() {
+		this(null, null);
+	}
+
+	public AttachCapabilitiesEvent(Class<T> type, T obj) {
+		super(type);
+		this.obj = obj;
+	}
+
+	/**
+	 * Retrieves the object that is being created, Not much state is set.
+	 */
+	public T getObject() {
+		return this.obj;
+	}
+
+	/**
+	 * Adds a capability to be attached to this object.
+	 * Keys MUST be unique, it is suggested that you set the domain to your mod ID.
+	 * If the capability is an instance of INBTSerializable, this key will be used when serializing this capability.
+	 *
+	 * @param key The name of owner of this capability provider.
+	 * @param cap The capability provider
```

Comment:
Maybe rename this to `provider`
```suggestion
	 * @param cap The {@link ICapabilityProvider}
```